feat: setup.py: remove rsa requirement

ktdreyer opened the pull request:
The `rsa` library is slower and not as well-maintained as the `cryptography` library. Now that we require the `cryptography` library, drop the hard requirement on the `rsa` library.


Fixes: #941

arithmetic1728 commented:
I think simply removing the rsa dependency will break the library since it is still in use: https://github.com/googleapis/google-auth-library-python/blob/main/google/auth/crypt/_python_rsa.py#L28

ktdreyer commented:
Please see https://github.com/googleapis/google-auth-library-python/issues/646 , the code does not use python-rsa if python-cryptography is installed.

ktdreyer commented:
This is still relevant because `pip install google-auth` still pulls in the old `rsa` library.

As I look at rebasing this today, I can remove the entry from `setup.py`, but it's now present in `.kokoro/requirements.txt` too. How do I make the change to that file?

ktdreyer commented:
I researched kokoro to understand how to remove `rsa` from `.kokoro/requirements.txt`.

`google-auth` requires `rsa`, so we need to remove `rsa` from `setup.py`, then push a new release:

```
google-auth
  └rsa
```

The complicating factor is that `.kokoro/requirements.in` lists two modules that are not in `setup.py`: (`gcp-docuploader` and `gcp-releasetool`). These entries cause `pip-compile` to fetch `google-auth=2.28.0` from PyPI. Because `2.28.0` still requires `rsa`, `pip-compile` still puts `rsa` into `requirements.txt`.

I think the solution here is to release a new `google-auth` version with this PR, then re-run `pip-compile`, and we should see the entry disappear from `.kokoro/requirements.txt`.

prodigysml commented:
Any chance we can get some eyes on this one? There's a dependabot alert we can't get rid of, that would be fixed with this PR.

miketheman commented:
This might need to be revived due to https://github.com/sybrenstuvel/python-rsa/issues/245

miketheman commented:
I _think_ the build failure can be resolved by adding the missing `rsa` package to
https://github.com/googleapis/google-auth-library-python/blob/56b4d3b3bfd6f935411130c8828cdeea90aa69f1/docs/conf.py#L375

Ref: https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_mock_imports

That way, when sphinx autodoc traverses:

`google.auth._jwt_async` -> `google.auth.jwt` -> `google.auth.crypt` -> `google.auth.crypt.rsa` ->  (sphinx follows fallback) `google.auth.crypt._python_rsa` -> `import rsa`

That final import to the missing `rsa` package will be ignored.

ktdreyer committed:
feat: setup.py: remove rsa requirement

The rsa library is slower and not as well-maintained as the cryptography
library. Now that we require the cryptography library, drop the hard
requirement on the rsa library.

ktdreyer commented:
@parthea would you please merge this? It simply drops `rsa` from the list of mandatory requirements.

daniel-sanche commented:
Thanks for putting this together and bringing it to our attention. I'll have to look into this a bit closer to make sure I understand the consequences of removing the dependency. Presumably, we'll need to add something else to provide a default implementation

di commented:
There is a polyfill here that already prefers the `pyca/cryptography` implementation and only falls back to `rsa` if it is unavailable, so in effect there already is a "default" implementation. 

https://github.com/googleapis/google-auth-library-python/blob/56b4d3b3bfd6f935411130c8828cdeea90aa69f1/google/auth/crypt/rsa.py#L18-L30

daniel-sanche commented:
Yeah, I think we're going to have to make cryptography a default dependency instead of an optional. Closing this in favor of https://github.com/googleapis/google-auth-library-python/pull/1771, which is making that change. Thanks!
